compiler/rustc_infer/src/infer/error_reporting/sub_relations.rs

```diff
@@ -0,0 +1,71 @@
+use rustc_data_structures::fx::FxHashMap;
+use rustc_data_structures::undo_log::NoUndo;
+use rustc_data_structures::unify as ut;
+use rustc_middle::ty;
+
+use crate::infer::InferCtxt;
+
+#[derive(Debug, Copy, Clone, PartialEq)]
+struct SubId(u32);
+impl ut::UnifyKey for SubId {
+    type Value = ();
+    #[inline]
+    fn index(&self) -> u32 {
+        self.0
+    }
+    #[inline]
+    fn from_index(i: u32) -> SubId {
+        SubId(i)
+    }
+    fn tag() -> &'static str {
+        "SubId"
+    }
+}
+
+/// When reporting ambiguity errors, we sometimes want to
+/// treat all inference vars which are subtypes of each
+/// others as if they are equal. For this case we compute
+/// the transitive closure of our subtype obligations here.
+#[derive(Default)]
+pub struct SubRelations {
```

Comment:
Question: do we have tests that exercise nontrivial sub relations that aren't just like... `?0 <: W<?0>`?

I'm just curious if this machinery is necessary for tests to be improved, or if we can special case the only types of subtyping errors that we actually see in practice.

I mostly ask this since this is implemented in the same PR as the one that removes, so I can't see the diagnostic implications of this.

Reply:
extended the comment

see https://github.com/rust-lang/rust/commit/5e8d8892e814fc82c9df890201a6e91b3c09e6fa for the impact on error messages